docs/rfcs/0004-bot-friendly-front-end.md

```diff
@@ -0,0 +1,96 @@
+## Making mitxpro bot-friendly
+
+## Abstract
+
+In Open Discussions we ran into an issue that is common for single page 
+apps in general. We wanted to add some `<meta>` tags to certain pages that
+would create an attractive preview when shared on Facebook (for an example, 
+look at the Link Preview section [here](https://developers.facebook.com/tools/debug/sharing/?q=http%3A%2F%2Fmit.edu)).
+The problem: Facebook has a crawler that scans for these particular `<meta>`
+tags to create that preview, but it doesn't execute any Javascript. As a result,
+the crawler could only see the basically-empty Django template for the page.
+
+In Open Discussions we ended up serving up a basic page with no content and
+some default `<meta>` tags for all requests from the Facebook crawler user agent. 
+In xpro, we will almost definitely want to support Facebook link previews, and
+we will want to have different previews for different pages (e.g.: on a course
+detail page, show the actual course image and course title rather than some default 
+image/text).  
+
+This RFC aims to lay out our options for achieving that, list some advantages and
+drawbacks, and gather feedback about which option to take.
+
+## Options
+
+#### 1) Serve specific content to the Facebook user agent (like we do in OD)
+
+With this approach, we would have some specialized logic for rendering a Django
+template if the request has the Facebook crawler user agent. Depending on the 
+route being requested, we would provide different bits of template data. For example,
+if the request was for the root URL, we would render the template with some default title
+value. If the request was for a specific course detail page, we would render the template
+with the course title as the title value.
+
+*Advantages:*
+1. Easiest and quickest option. Easy to implement alongside our current patterns and UI 'philosophy'.
+1. No new libraries/tools to learn.
+
+*Drawbacks:*
+1. Logic copying and the possibility of drift. With this approach we would have
+  two different sets of logic for rendering some routes, and it would be split
+  between the front- and back-end. It would also be possible for the front- and 
+  back-end logic to drift, i.e.: the front-end renders some content for a route that
+  doesn't match the Facebook link preview content for the same route.
+1. If any new pages/routes were created for which we would want some new
+  Facebook preview content, it would be very easy to forget to implement the 
+  specialized logic for the Facebook-UA-specific template. We'd be much
+  more likely to remember if the `<meta>` tag rendering was with the rest of
+  the React code (which would be the case if we used server-side rendering).
+1. It's possible that we'd want to add support for other bots (Twitter? Google?) in
+  the future. That would add some complexity to the template-rendering logic.
```

Comment:
couldn't we simply always include the `<meta>` tags in the django template we render, regardless of the user agent? that wouldn't address the issues you raise of possible drift between frontend and backend, but would make our implementation less facebook specific (and obviate the need to add specific support for any other services we want to support in the future).

Reply:
In open-discussions we wanted different Facebook preview content for different pages, e.g.: a card with an article cover image and post title for a post detail page, a pic of the MIT dome with a generic title for the home page, a channel header image and channel title for a channel page. Those meta tags can always be included in the template, but we can't give them the right page-specific-values without performing certain queries/etc. in Django and providing the values to the Django template

Reply:
ok that makes sense as to why there needs to be route-specific logic, carry on!